File: libmuscle/python/libmuscle/peer_manager.py
from typing import cast, Dict, List

from ymmsl import Conduit, Identifier, Reference

from libmuscle.endpoint import Endpoint


class PeerManager:
    """Manages information about peers for a Communicator
    """
    def __init__(self, kernel: Reference, index: List[int],
                 conduits: List[Conduit],
                 peer_dims: Dict[Reference, List[int]],
                 peer_locations: Dict[Reference, List[str]]) -> None:
        """Create a PeerManager.

        Peers here are instances, and peer_dims and peer_locations are
        indexed by a Reference to an instance. Instance sets are
        multi-dimensional arrays with sizes given by peer_dims.

        Args:
            kernel: The kernel for the instance whose peers we're
                    managing.
            index: The index of the instance whose peers we're
                    managing.
            conduits: A list of conduits attached to this compute
                    element, as received from the manager.
            peer_dims: For each peer we share a conduit with, the
                    dimensions of the instance set.
            peer_locations: A list of locations for each peer instance
                    we share a conduit with.
        """
        self.__kernel = kernel
        self.__index = index

        # peer port ids, indexed by local kernel.port id
        self.__peers = dict()  # type: Dict[Reference, Reference]

        for conduit in conduits:
            if str(conduit.sending_component()) == str(kernel):
                # we send on the port this conduit attaches to
                self.__peers[conduit.sender] = conduit.receiver
            if str(conduit.receiving_component()) == str(kernel):
                # we receive on the port this conduit attaches to
                self.__peers[conduit.receiver] = conduit.sender

        self.__peer_dims = peer_dims    # indexed by kernel id
        self.__peer_locations = peer_locations  # indexed by instance id

    def is_connected(self, port: Identifier) -> bool:
        """Determine whether the given port is connected.

        Args:
            port: The port to check.
        """
        recv_port_full = self.__kernel + port
        return recv_port_full in self.__peers

    def get_peer_port(self, port: Identifier) -> Reference:
        """Get a reference for the peer port.

        Args:
            port: Name of the port on this side.
        """
        return self.__peers[self.__kernel + port]

    def get_peer_dims(self, peer_kernel: Reference) -> List[int]:
        """Get the dimensions of a peer kernel.

        Args:
            peer_kernel: The peer kernel whose dimensions to get.
        """
        return self.__peer_dims[peer_kernel]

    def get_peer_locations(self, peer_instance: Reference) -> List[str]:
        """Get the locations of a peer instance.

        There may be multiple, if the peer supports more than one
                protocol.

        Args:
            peer_instance: The instance whose locations to get.
        """
        return self.__peer_locations[peer_instance]

    def get_peer_endpoint(self, port: Identifier, slot: List[int]
                          ) -> Endpoint:
        """Determine the peer endpoint for the given port and slot.

        Args:
            port: The port on our side to send or receive on.
            slot: The slot to send or receive on.

        Returns:
            The peer endpoint.
        """
        peer = self.__peers[self.__kernel + port]
        peer_kernel = cast(Reference, peer[:-1])
        peer_port = cast(Identifier, peer[-1])

        total_index = self.__index + slot

        # rebalance the indices
        peer_dim = len(self.__peer_dims[peer_kernel])
        peer_index = total_index[0:peer_dim]
        peer_slot = total_index[peer_dim:]
        return Endpoint(peer_kernel, peer_index, peer_port, peer_slot)

File: libmuscle/cpp/src/libmuscle/peer_manager.cpp
#include <libmuscle/peer_manager.hpp>

#include <sstream>

using ymmsl::Conduit;
using ymmsl::Identifier;
using ymmsl::Reference;


namespace libmuscle { namespace _MUSCLE_IMPL_NS {

PeerManager::PeerManager(
        Reference const & kernel,
        std::vector<int> const & index,
        std::vector<Conduit> const & conduits,
        PeerDims const & peer_dims,
        PeerLocations const & peer_locations
        )
    : kernel_(kernel)
    , index_(index)
    , peers_()                          // peer port ids, indexed by local kernel.port
    , peer_dims_(peer_dims)             // indexed by peer kernel id
    , peer_locations_(peer_locations)   // indexed by peer instance id
{
    for (auto const & conduit : conduits) {
        if (conduit.sending_component() == kernel_) {
            // we send on the port this conduit attaches to
            auto search = peers_.find(conduit.sender);
            if (search == peers_.end())
                search = peers_.emplace(
                        conduit.sender, std::vector<Reference>()).first;
            search->second.push_back(conduit.receiver);
        }
        if (conduit.receiving_component() == kernel_) {
            // we receive on the port this conduit attaches to
            if (peers_.count(conduit.receiver)) {
                std::stringstream ss;
                ss << "Receiving port \"" << conduit.receiving_port();
                ss << "\" is connected by multiple conduits, but at most one";
                ss << " is allowed.";
                throw std::runtime_error(ss.str());
            }
            std::vector<Reference> vec = {conduit.sender};
            peers_.emplace(conduit.receiver, vec);
        }
    }
}

bool PeerManager::is_connected(Identifier const & port) const {
    return peers_.count(kernel_ + port);
}

 std::vector<ymmsl::Reference> const & PeerManager::get_peer_ports(
            Identifier const & port) const {
    return peers_.at(kernel_ + port);
}

std::vector<int> PeerManager::get_peer_dims(
        Reference const & peer_kernel) const {
    return peer_dims_.at(peer_kernel);
}

std::vector<std::string> PeerManager::get_peer_locations(
        Reference const & peer_instance) const {
    return peer_locations_.at(peer_instance);
}

std::vector<Endpoint> PeerManager::get_peer_endpoints(
        Identifier const & port,
        std::vector<int> const & slot
        ) const
{
    auto peers = peers_.at(kernel_ + port);
    std::vector<Endpoint> endpoints;

    for (auto peer : peers) {
        Reference peer_kernel(peer.cbegin(), std::prev(peer.cend()));
        Identifier peer_port = std::prev(peer.cend())->identifier();

        std::vector<int> total_index = index_;
        total_index.insert(total_index.end(), slot.cbegin(), slot.cend());

        // rebalance the indices
        int peer_dim = peer_dims_.at(peer_kernel).size();
        auto peer_dim_it = std::next(total_index.cbegin(), peer_dim);
        std::vector<int> peer_index(total_index.cbegin(), peer_dim_it);
        std::vector<int> peer_slot(peer_dim_it, total_index.cend());
        endpoints.emplace_back(peer_kernel, peer_index, peer_port, peer_slot);
    }

    return endpoints;
}

} }

